Replace hardcoded "Loading..." text with skeleton loaders

harsh04044 opened the pull request:
## Proposed Changes

Fixes #14704 

- **Replaced hardcoded "Loading..." text with skeleton loaders** in:
  - `LocationForm.tsx` - Uses `FormSkeleton` component for better visual feedback
  - `ResourceCategoryPicker.tsx` - Uses `Skeleton` component with loading spinner
  - `UpdateSpecimenDefinition.tsx` - Uses `FormSkeleton` component

- **Added accessibility support** to skeleton loaders:
  - Added `role="status"` and `aria-live="polite"` for screen reader announcements
  - Added `aria-label` with loading translation
  - Added visually hidden `sr-only` span for assistive technologies
  - Ensures loading states are properly announced to screen readers

- **Updated Playwright tests** to wait for skeleton loaders to disappear:
  - Added wait conditions in `locationCreation.spec.ts`
  - Added wait conditions in `locationEdit.spec.ts`
  - Added wait conditions in `specimenDefinitionEdit.spec.ts`

**Benefits:**
- Better UX: Skeleton loaders provide visual feedback about content structure while loading
- Improved accessibility: Screen readers can properly announce loading states
- Consistent loading patterns across the application
- Better test reliability with proper wait conditions

**Screenshots:**
<img width="2940" height="1912" alt="image" src="https://github.com/user-attachments/assets/5f9b5802-aae5-45bd-a548-961d14dddec8" />


Tagging: @ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate the bug or test the new feature. _(N/A - UX improvement, existing functionality works as expected)_
- [ ] Update [product documentation](https://docs.ohc.network). _(N/A - Internal UI improvement, no user-facing feature changes)_
- [x] Ensure that UI text is placed in I18n files. _(Loading text uses t("loading") translation)_
- [ ] Prepare a screenshot or demo video for the changelog entry and attach it to the issue. _(Optional - Visual improvement)_
- [x] Request peer reviews.
- [ ] Complete QA on mobile devices. _(N/A - Loading state improvement, no mobile-specific changes)_
- [x] Complete QA on desktop devices.
- [x] Add or update Playwright tests for related changes. _(Updated tests to handle skeleton loaders)_

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Replaced plain "Loading..." text with consistent skeleton placeholders across resource selection, location forms, and specimen definition screens for clearer loading feedback.

* **Accessibility**
  * Improved loading-state accessibility by adding ARIA live regions, status roles, and descriptive labels so assistive technologies announce progress more reliably.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

harsh04044 committed:
fix: replace hardcoded Loading text with t(loading) translation

harsh04044 committed:
Merge branch 'develop' into fix/use-loading-translation

harsh04044 committed:
Merge branch 'develop' into fix/use-loading-translation

harsh04044 committed:
refactor: use skeleton loaders instead of loading text

abhimanyurajeesh reviewed (changes requested):
<img width="967" height="586" alt="Image" src="https://github.com/user-attachments/assets/2801c377-e7c3-46e2-81e1-e4363bfff420" />

PR context and file changes are not matching, if you are using Skeleton please look into the accessibility issue too

harsh04044 commented on src/pages/Facility/settings/locations/LocationForm.tsx:
```diff
@@ -28,6 +28,8 @@ import {
 } from "@/components/ui/select";
 import { Textarea } from "@/components/ui/textarea";
 
+import { FormSkeleton } from "@/components/Common/SkeletonLoading";
```
i will addres it in next commit

harsh04044 commented on src/pages/Facility/settings/specimen-definitions/UpdateSpecimenDefinition.tsx:
```diff
@@ -3,6 +3,8 @@ import { useNavigate } from "raviger";
 import { useTranslation } from "react-i18next";
 import { toast } from "sonner";
 
+import { FormSkeleton } from "@/components/Common/SkeletonLoading";
```
this also

harsh04044 committed:
fix: add accessibility support to skeleton loaders

- Add role='status' and aria-live='polite' for screen reader announcements
- Add aria-label with loading translation
- Add visually hidden sr-only span for assistive technologies
- Addresses CodeRabbit accessibility suggestions

Fixes accessibility issues in:
- LocationForm.tsx
- UpdateSpecimenDefinition.tsx
- ResourceCategoryPicker.tsx

harsh04044 committed:
test: add wait conditions for form fields after skeleton loader changes

- Add explicit waits for form fields to appear after clicking edit buttons
- Fixes test failures caused by skeleton loaders replacing 'Loading...' text
- Tests now wait for skeleton loader to disappear before interacting with forms

Jacobjeevan commented on src/components/Common/ResourceCategoryPicker.tsx:
```diff
@@ -190,9 +190,15 @@ export function ResourceCategoryPicker({
+          aria-live="polite"
+          aria-label={t("loading")}
+        >
+          <span className="sr-only">{t("loading")}</span>
```
aria-label already handles accessibility, so this span is redundant. Same with below.

harsh04044 replied:
@Jacobjeevan i have addresed this in my latest commit

harsh04044 committed:
fix: remove redundant sr-only spans where aria-label handles accessibility

- Remove redundant sr-only spans from ResourceCategoryPicker, LocationForm, and UpdateSpecimenDefinition
- aria-label already provides accessibility, making sr-only spans unnecessary
- Addresses maintainer feedback on PR #14881

Jacobjeevan commented on src/components/Common/ResourceCategoryPicker.tsx:
```diff
@@ -190,9 +190,14 @@ export function ResourceCategoryPicker({
+          aria-live="polite"
+          aria-label={t("loading")}
+        >
           <Loader2 className="h-4 w-4 animate-spin" />
```
Thought I had mentioned this earlier, but must have missed it. Is this loader needed now? Skeleton alone should be enough, no?

harsh04044 replied:
You're  right I've removed the Loader2 component - the Skeleton alone is sufficient for the loading state. Thanks for catching that!

harsh04044 committed:
refactor: remove redundant Loader2 from ResourceCategoryPicker loading state

The Skeleton component alone is sufficient for indicating loading state.
Removed the redundant Loader2 spinner as suggested by maintainer.

abhimanyurajeesh commented on tests/facility/settings/locations/locationCreation.spec.ts:
```diff
@@ -66,6 +66,11 @@ test.describe("Facility Location Creation", () => {
+      await expect(
+        page.getByRole("combobox", { name: "Location Form" }),
+      ).toBeVisible({ timeout: 10000 });
+
```
why?

harsh04044 replied:
You're right  , we don't need to check for skeleton loaders. toBeVisible() already waits for the element, which implies any skeleton loaders have disappeared. I'll remove that comment and keep the wait for the form field.

abhimanyurajeesh reviewed (changes requested):
do we really need to look for skeletons in test ?

cc: @nihal467

yash-learner commented:
FYI @harsh04044 

For `AcitvityDefintionForm` it being done in https://github.com/ohcnetwork/care_fe/pull/15035

The reason how it happened https://github.com/ohcnetwork/care_fe/pull/15035#issuecomment-3717428937

So ignore that(AcitvityDefintionForm) in this PR, as it not done in this PR, as of now

cc: @abhimanyurajeesh @Jacobjeevan

harsh04044 committed:
refactor: remove explicit skeleton loader waits from tests

- Remove unnecessary explicit waits for skeleton loaders to disappear
- Playwright's toBeVisible() already handles waiting for elements
- Addresses maintainer feedback on PR #14881

abhimanyurajeesh committed:
Merge branch 'develop' into fix/use-loading-translation

harsh04044 commented on src/pages/Facility/settings/locations/LocationForm.tsx:
```diff
@@ -263,7 +265,16 @@ export default function LocationForm({
+    return (
+      <div
+        className="p-4"
+        role="status"
+        aria-live="polite"
+        aria-label={t("loading")}
+      >
+        <FormSkeleton rows={6} />
+      </div>
+    );
```
i am fixing this in my upcoming commit

harsh04044 committed:
fix: replace div with output element in LocationForm loading state

- Replace wrapping div with semantic output element
- Add sr-only span for screen reader accessibility
- Keep FormSkeleton unchanged
- Satisfies lint/a11y/useSemanticElements rule

harsh04044 committed:
fix: remove redundant aria attributes from LocationForm loading state

- Remove aria-live and aria-label from output element
- Keep only sr-only span for single accessibility announcement
- Prevents duplicate loading announcements for screen readers

harsh04044 commented on src/pages/Facility/settings/locations/LocationForm.tsx:
```diff
@@ -263,7 +265,12 @@ export default function LocationForm({
   ];
 
   if (locationId && isLoading) {
-    return <div className="p-4">Loading...</div>;
+    return (
+      <output className="p-4" aria-live="polite" aria-label={t("loading")}>
+        <span className="sr-only">{t("loading")}</span>
+        <FormSkeleton rows={6} />
+      </output>
+    );
   }
```
@coderabbitai i have made this change

harsh04044 committed:
fix: replace output with div and add ARIA status attributes in LocationForm

- Replace non-semantic output element with neutral div container
- Add role="status", aria-live="polite", and aria-label for accessibility
- Keep sr-only span for screen reader announcement
- Match accessibility pattern used in ResourceCategoryPicker and UpdateSpecimenDefinition

harsh04044 committed:
fix: replace div with semantic output element in LocationForm loading state

- Replace div with semantic output element (implicit status role)
- Remove redundant role="status" attribute
- Remove aria-label to prevent duplicate screen reader announcements
- Keep aria-live="polite" and sr-only span for accessibility
- Maintain FormSkeleton component unchanged
